### ISSUE TITLE
The dashboard page results in very high web memory consumption node when the number jobs is very large 

### ISSUE THREAD

**Issue by xtreme-sameer-vohra** [concourse/concourse#5341]
## Bug Report
We observe very high memory consumption when attempting to open the dashboard page to the point where multiple browser sessions to the the dashboard page caused the web to run out of memory on a 16gb VM.

Number of pipelines: 5000
Number of jobs: 36000

Only 10 pipelines were running, everything else was paused. Looking at the heap dumps, we were able to observe that 90+ % of the memory was being consumed in order to service the ListAllJobs endpoint ( see attached screen ).

The CPU consumption was high as well.

## Expected Results

The Dashboard page doesn't cause the web node to fall over.

## Actual Results

The web node memory spiked and was killed.

![Screen Shot 2020-03-20 at 1 20 46 PM](https://user-images.githubusercontent.com/1062564/77194303-b6623f80-6aad-11ea-8a6a-2037f2ef3b16.png)

runtime.Memstats -> `curl http://localhost:8079/debug/pprof/heap?debug=1 > DUMP_FILE`
```
# runtime.MemStats
# Alloc = 5063174048
# TotalAlloc = 116274766592
# Sys = 18161311632
# Lookups = 0
# Mallocs = 1070905432
# Frees = 1035748650
# HeapAlloc = 5063174048
# HeapSys = 17243832320
# HeapIdle = 9128198144
# HeapInuse = 8115634176
# HeapReleased = 638550016
# HeapObjects = 35156782
# Stack = 3145728 / 3145728
# MSpan = 141708192 / 198393856
# MCache = 13888 / 16384
# BuckHashSys = 2737420
# GCSys = 669620224
# OtherSys = 43565700
# NextGC = 6126008000
# LastGC = 1584656981086041776
```
The Alloc & HeapInUse account for 13 GB ( ~90% of the available memory ).

## Version Info

* Concourse version: v5.2.5
* Deployment type (BOSH/Docker/binary): BOSH ( however, this would affect all deployment types )
* Infrastructure/IaaS: AWS
* Browser (if applicable): N/A
* Did this used to work? No

**Comment by pivotal-jamie-klassen** [comment]
Still needs further investigation, but the working theory is that #5336 should address this for 5.2.x and #5340 should address this for 5.5.x.

**Comment by cirocosta** [comment]
Hey,

I was doing some experiments here to have a clear view of how this looks like on
the 5.2.x series.

It indeed doesn't look good :grimacing:

![Screen Shot 2020-03-23 at 11 10 09-fullpage](https://user-images.githubusercontent.com/3574444/77332420-6f1bbf00-6cf8-11ea-968d-c683016f1744.png)


That's from a setup consisting of

- 1800 pipelines
- 30000 jobs
- 90000 builds

hitting `api/v1/jobs` (logged in) with a varying number of concurrent requests
(from 1 to 8).



### reproducing


1. start a deployment of Concourse with the database preloaded with those
pipelines

you can do so by bringing concourse up (from the `release/5.2.x` branch) using
the following `docker-compose.yml`:

```yaml
version: '3'

services:
  db:
    image: cirocosta/postgres:concourse-data
    ports:
    - 6543:5432
    environment:
      PGDATA: "/data"
      POSTGRES_DB: concourse
      POSTGRES_USER: dev
      POSTGRES_PASSWORD: dev

  web:
    build: .
    image: concourse/concourse:local
    volumes:
    - .:/src
    command: web
    depends_on: [db]
    ports:
    - 8080:8080
    - 8079:8079
    environment:
      CONCOURSE_DEBUG_BIND_IP: "0.0.0.0"
      CONCOURSE_PROMETHEUS_BIND_IP: "0.0.0.0"
      CONCOURSE_PROMETHEUS_BIND_PORT: "9100"
      CONCOURSE_LOG_LEVEL: debug
      CONCOURSE_POSTGRES_HOST: db
      CONCOURSE_POSTGRES_USER: dev
      CONCOURSE_POSTGRES_PASSWORD: dev
      CONCOURSE_POSTGRES_DATABASE: concourse
      CONCOURSE_EXTERNAL_URL: http://localhost:8080
      CONCOURSE_ADD_LOCAL_USER: test:test,guest:guest
      CONCOURSE_MAIN_TEAM_LOCAL_USER: test
      CONCOURSE_CLUSTER_NAME: dev

  prometheus:
    image: prom/prometheus
    entrypoint:
      - /bin/sh
      - -c
      - |
        echo "
        global:
          scrape_interval: '5s'
          evaluation_interval: '5s'
        scrape_configs:
          - job_name: 'concourse'
            static_configs:
              - targets:
                - 'web:9100'
        " > config.yml
          exec prometheus \
            --config.file=config.yml \
            --storage.tsdb.path=/prometheus
    ports:
      - '9090:9090'
```

that `db` service is essentially `postgres` but without


2. hit the `/api/v1/jobs` endpoints w/ varying concurrency

```console
# log in
#
$ fly -t local login -u test -p test -c http://localhost:8080

# trigger 10 concurrent requests
#
$ for i in $(seq 1 10); do fly -t local curl /api/v1/jobs & done
```

**Comment by cirocosta** [comment]
Here's the `alloc_space` (total that has been allocated over all of the runs - hitting this endpoint 44 times)

![profile001](https://user-images.githubusercontent.com/3574444/77333552-fc134800-6cf9-11ea-9493-3a0b160cc831.png)

**Comment by cirocosta** [comment]
Hey,

Here's another set of datapoint on how the v5.x.x series compares against
v6.0.0.

To get a sense of how memory usage scales with regards to the number of jobs
I set up the following experiment:

- starting with 30k jobs, remove 1k jobs on each iteration, making a request to
  /api/v1/jobs and capturing the "how big is the heap right now" information

- repeat this 20 times (so we can better deal with any possible variance)

here are the results:

![v5 5 x](https://user-images.githubusercontent.com/3574444/77494272-b2228300-6e1b-11ea-958b-38413aecc973.png)

![v6 0 0](https://user-images.githubusercontent.com/3574444/77494274-b2bb1980-6e1b-11ea-9b88-0cfd83cffad8.png)

note how different the scale is - while in the v5.5.x series a request can make
the web node take even up to 1GB (with 30k visible jobs), we don't pass the
60MB mark for the 6.x.x series.

coupled with the other issues I linked to above, v6.0.0 should be huuge for
those suffering with dashboard right now.

thanks!
